### PR TITLE
Add a Rust timestamp converter transform

### PR DESCRIPTION
**Pull request by voutilad**
This transform is based off Confluent's SMT for timestamp conversion as part of Kafka Connect. The initial version uses Schema Registry and supports Avro for converting timestamp values in keys or values, including:

- field-based conversion of Avro Records (e.g. convert just a field in a Record)
- conversion from most Avro Value types to reasonable default outputs
- parsing Avro String input as well as conversion to Avro String outputs

It's a pretty comprehensive Rust transform example using the new Schema Registry components that are still cooking. Includes lots of unit tests as well as an integration test that stubs out things like the SR client and a record writer to test the main Avro conversion function just like it's called by the wasm runtime.

### CODE REVIEW

**Comment by voutilad** [comment]
Failed to mention, this is blocked by a few bug fixes landing in the public Rust crates. Until then, it requires building with the Redpanda rust-sdk source locally available. I fully expect the doc tests to fail until then 😬

**Comment by voutilad** [comment]
This transform needs this to be merged and new crates to be published before it can drop any local redpanda dependencies. https://github.com/redpanda-data/redpanda/pull/18555/files

**Comment by voutilad** [comment]
Ok, now that some of the Rust SDK changes are live, this PR no long requires building with a local copy of Redpanda.

File: data-transforms/ts-converter/README.adoc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+- At least version 1.75 of https://rustup.rs/[Rust^] installed on your host machine.
+- The Wasm target for Rust installed. To install this target, run the following:
++
+[source,bash]
+----
+rustup target add wasm32-wasi
+----
+// (step {"action":"runShell", "command": "rustup target add wasm32-wasi"})
+- link:https://docs.redpanda.com/current/get-started/rpk-install/[Install `rpk`] on your host machine.
+- https://docs.docker.com/compose/install/[Docker and Docker Compose] installed on your host machine.
```
Comment (JakeSCahill) [inline data-transforms/ts-converter/README.adoc:37]:
```suggestion
- https://docs.docker.com/compose/install/[Docker and Docker Compose^] installed on your host machine.
```

This notation makes the link open a new tab. It's standard in our docs to do this for external links.

File: data-transforms/ts-converter/README.adoc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+```
+
+. Change into the `data-transforms/ts-converter/` directory:
++
+[,bash]
+----
+cd redpanda-labs/data-transforms/ts-converter
+----
+
+. Set the `REDPANDA_VERSION` environment variable to at least version 24.1.2. For all available versions, see the https://github.com/redpanda-data/redpanda/releases[GitHub releases].
```
Comment (JakeSCahill) [inline data-transforms/ts-converter/README.adoc:54]:
```suggestion
. Set the `REDPANDA_VERSION` environment variable to at least version 24.1.2. For all available versions, see the https://github.com/redpanda-data/redpanda/releases[GitHub releases^].
```

File: data-transforms/ts-converter/README.adoc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
++
+For example:
++
+[,bash,subs="attributes+"]
+----
+export REDPANDA_VERSION={latest-redpanda-version}
+----
+
+. Set the `REDPANDA_CONSOLE_VERSION` environment variable to the version of Redpanda Console that you want to run. For all available versions, see the https://github.com/redpanda-data/redpanda/releases[GitHub releases].
```
Comment (JakeSCahill) [inline data-transforms/ts-converter/README.adoc:63]:
```suggestion
. Set the `REDPANDA_CONSOLE_VERSION` environment variable to the version of Redpanda Console that you want to run. For all available versions, see the https://github.com/redpanda-data/redpanda/releases[GitHub releases^].
```

File: data-transforms/ts-converter/README.adoc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+```
+// (step {"action":"runShell", "command": "rpk transform build"})
+// (step {"action":"runShell", "command": "bash ./doc-tests/deploy-transform.sh"})
+// Wait for the transform to be processed by Redpanda
+// (step {"action":"wait", "duration": 10000})
++
+This example accepts the following environment variables:
++
+- `TIMESTAMP_TARGET_TYPE` (*required*): The output conversion time. Must be one of:
+* `string[<pattern>]`: where `<pattern>` is a valid https://docs.rs/chrono/latest/chrono/format/strftime/index.html[chrono format] string
```
Comment (JakeSCahill) [inline data-transforms/ts-converter/README.adoc:118]:
```suggestion
* `string[<pattern>]`: where `<pattern>` is a valid https://docs.rs/chrono/latest/chrono/format/strftime/index.html[chrono format^] string
```

File: data-transforms/ts-converter/README.adoc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+{
+  "topic": "sink",
+  "value": "\"2024-01-03T19:38:06.988+00:00\"",
+  "timestamp": 1715890281087,
+  "partition": 0,
+  "offset": 0
+}
+----
+
+You can also see this in http://localhost:8080/topics/sink?p=-1&s=50&o=-1#messages[Redpanda Console].
```
Comment (JakeSCahill) [inline data-transforms/ts-converter/README.adoc:152]:
```suggestion
You can also see this in http://localhost:8080/topics/sink?p=-1&s=50&o=-1#messages[Redpanda Console^].
```

**Review by JakeSCahill — COMMENTED** [review]
Looks great! Thanks, Dave. I tested this and it worked but requires one change. I also left some content suggestions.

If you want this to be tested as part of our CI pipeline, you need to add it as a step here: https://github.com/redpanda-data/redpanda-labs/blob/main/.github/workflows/test-docs.yml

File: .github/workflows/test-docs.yml [hunk 6]
```diff
@@ -15,6 +15,7 @@ jobs:
       data-transforms-redaction: ${{ steps.filter.outputs.data-transforms-redaction }}
       data-transforms-regex: ${{ steps.filter.outputs.data-transforms-regex }}
       data-transforms-jq: ${{ steps.filter.outputs.data-transforms-jq }}
+      data-transforms-ts-converter: ${{ steps.filter.outputs.data-tramsforms-ts-converter }}
```
Comment (JakeSCahill) [inline .github/workflows/test-docs.yml:18]:
```suggestion
      data-transforms-ts-converter: ${{ steps.filter.outputs.data-transforms-ts-converter }}
```

**Pull request merged**